ironfish/src/wallet/scanner/walletScanner.ts

```diff
@@ -252,21 +254,48 @@ export class WalletScanner {
 
     throw new Error('WalletScanner requires either chain or client')
   }
-}
-
-function getTransactionsWithNoteIndex(
-  header: BlockHeader,
-  transactions: Transaction[],
-): Array<{ transaction: Transaction; initialNoteIndex: number }> {
-  Assert.isNotNull(header.noteSize)
-  let initialNoteIndex = header.noteSize
 
-  const result = []
+  /**
+   * Checks whether `scanningAccounts` is stale or up-to-date.
+   */
+  private haveWalletAccountsChanged(): boolean {
+    const accountIds = new Set(
+      this.wallet
+        .listAccounts()
+        .filter((account) => account.scanningEnabled)
+        .map((account) => account.id),
+    )
+    return (
+      this.scanningAccounts.length !== accountIds.size ||
+      !this.scanningAccounts.every(({ account }) => accountIds.has(account.id))
+    )
+  }
 
-  for (const transaction of transactions.slice().reverse()) {
-    initialNoteIndex -= transaction.notes.length
-    result.push({ transaction, initialNoteIndex })
+  /**
+   * Replaces `scanningAccounts` with fresh values from the wallet.
+   */
+  private async refreshScanningAccounts(): Promise<void> {
+    this.scanningAccounts = await Promise.all(
+      this.wallet
+        .listAccounts()
+        .filter((account) => account.scanningEnabled)
+        .map(async (account) => ({
+          account,
+          scanFrom: await account.getHead(),
+        })),
+    )
   }
 
-  return result.slice().reverse()
+  private getEarliestHead(): HeadValue | null {
+    let earliestHead = null
+    for (const { scanFrom: head } of this.scanningAccounts) {
+      if (!head) {
+        return null
```

Comment:
Note to look into this: but a null head just means the head is the block before the genesis block. Wouldn't this be considered the earliest head? I need to read the rest of the code to see how this is handled now.